modules/table/src/test/java/org/apache/ignite/internal/table/distributed/replicator/SchemaCompatibilityValidatorTest.java

```diff
@@ -0,0 +1,655 @@
+        return new CatalogTableColumnDescriptor(columnName, INT32, true, DEFAULT_PRECISION, DEFAULT_SCALE, DEFAULT_LENGTH, null);
+    }
+
+    private static CatalogTableColumnDescriptor intColumnWithDefault(String columnName, int defaultValue) {
+        return new CatalogTableColumnDescriptor(
+                columnName,
+                INT32,
+                false,
+                DEFAULT_PRECISION,
+                DEFAULT_SCALE,
+                DEFAULT_LENGTH,
+                DefaultValue.constant(defaultValue)
+        );
+    }
+
+    private static FullTableSchema tableSchema(int schemaVersion, List<CatalogTableColumnDescriptor> columns) {
+        return tableSchema(schemaVersion, TABLE_NAME, columns);
+    }
+
+    private static FullTableSchema tableSchema(int schemaVersion, String name, List<CatalogTableColumnDescriptor> columns) {
+        return new FullTableSchema(schemaVersion, TABLE_ID, name, columns);
+    }
+
+    @FunctionalInterface
+    private interface SchemaChangeSource {
+        List<FullTableSchema> schemaVersions();
+    }
+
+    private enum ForwardCompatibleChange implements SchemaChangeSource {
+        ADD_NULLABLE_COLUMN(() -> List.of(
+                tableSchema(1, List.of(
+                        intColumn("col1")
+                )),
+                tableSchema(2, List.of(
+                        intColumn("col1"),
+                        nullableIntColumn("col2")
+                ))
+        )),
+        ADD_COLUMN_WITH_DEFAULT(() -> List.of(
+                tableSchema(1, List.of(
+                        intColumn("col1")
+                )),
+                tableSchema(2, List.of(
+                        intColumn("col1"),
+                        intColumnWithDefault("col2", 42)
+                ))
+        )),
+        // TODO: https://issues.apache.org/jira/browse/IGNITE-20948 - uncomment this.
+        //RENAME_COLUMN(() -> List.of(
+        //        tableSchema(1, List.of(
+        //                intColumn("col1")
+        //        )),
+        //        tableSchema(2, List.of(
+        //                intColumn("col2")
+        //        ))
+        //)),
+        DROP_NOT_NULL(() -> List.of(
+                tableSchema(1, List.of(
+                        intColumn("col1")
+                )),
+                tableSchema(2, List.of(
+                        nullableIntColumn("col1")
+                ))
+        ));
+
+        private final Supplier<List<FullTableSchema>> schemaVersionsSupplier;
+
+        ForwardCompatibleChange(Supplier<List<FullTableSchema>> schemaVersionsSupplier) {
+            this.schemaVersionsSupplier = schemaVersionsSupplier;
+        }
+
+        @Override
+        public List<FullTableSchema> schemaVersions() {
+            return schemaVersionsSupplier.get();
+        }
+    }
+
+    private enum ForwardIncompatibleChange implements SchemaChangeSource {
+        RENAME_TABLE(() -> List.of(
```

Comment:
What is the reason of passing a supplier instead of passing the list itself? A memory constraint, or something different?